Remaining regressions seem to be due to `_assumeNonnegative(count)` not being enough -- the `count >= 0` precondition evidently enables optimizations that this is currently missing. (FWIW, `_uncheckedUnsafeAssume` slows down `FlattenListFlatMap` by another factor of 2.) Or for all I know it could just be an inlining instability -- I have no idea what I'm doing. ;-) (The weird part is that the SIL/disassembly looks the same before/after; the difference may be hiding in one of the stdlib dylib entry points.)

The macOS test regression is another case of `utf8_decoding_fastpath.swift`:

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-main/swift/test/SILOptimizer/utf8_decoding_fastpath.swift:84:15: error: CHECK-DAG: expected string not found in input
// CHECK-DAG: function_ref {{.*}}_fromUTF8Repairing
              ^
<stdin>:690:84: note: scanning from here
sil @$s22utf8_decoding_fastpath022decodeStringUTF8ViewAsF0yS2S0fG0VF : $@convention(thin) (@guaranteed String.UTF8View) -> @owned String {
                                                                                   ^
<stdin>:698:7: note: possible intended match here
 %5 = function_ref @$sSS8UTF8ViewV32withContiguousStorageIfAvailableyxSgxSRys5UInt8VGKXEKlFSS_Tg50104$sSRy7ElementSTQzGSSs5Error_pIgyozo_ACSSsAD_pIegyrzo_SlRzs16_UnicodeEncodingR_8CodeUnitQy_AARtzr0_lTRSS8ab42V_s0C0O0G0OTG5033$sSS8decoding2asSSx_q_mtcst10_cd3R_8ef4Y58_7a26Rtzr0_lufcSSSRyAGGXEfU_SS8gH14V_s0C0O0H0OTg5Tf3nnpf_nTf1cn_n : $@convention(thin) (@guaranteed String.UTF8View) -> (@owned Optional<String>, @error Error) // user: %7
      ^
```

The tiny addition of the `_assumeNonnegative` call made the compiler decide not to inline either decoding closure this time, which is Just Great. This time I'm resolving it by adding an alternate UBP initializer, like with `Range.init(_uncheckedBounds:)`.

--- a/stdlib/public/core/ClosedRange.swift
+++ b/stdlib/public/core/ClosedRange.swift
@@ -68,6 +68,8 @@ public struct ClosedRange<Bound: Comparable> {
   /// The range's upper bound.
   public let upperBound: Bound
 
+  // This works around _debugPrecondition() impacting the performance of
+  // optimized code. (rdar://72246338)
   @_alwaysEmitIntoClient @inline(__always)
   internal init(_uncheckedBounds bounds: (lower: Bound, upper: Bound)) {
     self.lowerBound = bounds.lower

--- a/stdlib/public/core/Range.swift
+++ b/stdlib/public/core/Range.swift
@@ -157,6 +157,8 @@ public struct Range<Bound: Comparable> {
   /// instance does not contain its upper bound.
   public let upperBound: Bound
 
+  // This works around _debugPrecondition() impacting the performance of
+  // optimized code. (rdar://72246338)
   @_alwaysEmitIntoClient @inline(__always)
   internal init(_uncheckedBounds bounds: (lower: Bound, upper: Bound)) {
     self.lowerBound = bounds.lower

--- a/stdlib/public/core/SmallString.swift
+++ b/stdlib/public/core/SmallString.swift
@@ -208,7 +208,7 @@ extension _SmallString {
         // Restore the memory type of self._storage
         _ = rawPtr.bindMemory(to: RawBitPattern.self, capacity: 1)
       }
-      return try f(UnsafeBufferPointer(start: ptr, count: count))
+      return try f(UnsafeBufferPointer(_uncheckedStart: ptr, count: count))
     }
   }
 

--- a/stdlib/public/core/StringObject.swift
+++ b/stdlib/public/core/StringObject.swift
@@ -902,7 +902,7 @@ extension _StringObject {
       return sharedUTF8
     }
     return UnsafeBufferPointer(
-      start: self.nativeUTF8Start, count: self.largeCount)
+      _uncheckedStart: self.nativeUTF8Start, count: self.largeCount)
   }
 
   // Whether the object stored can be bridged directly as a NSString

--- a/stdlib/public/core/UnsafeBufferPointer.swift.gyb
+++ b/stdlib/public/core/UnsafeBufferPointer.swift.gyb
@@ -394,6 +394,15 @@ extension Unsafe${Mutable}BufferPointer: ${Mutable}Collection, RandomAccessColle
 }
 
 extension Unsafe${Mutable}BufferPointer {
+  @_alwaysEmitIntoClient
+  internal init(
+    @_nonEphemeral _uncheckedStart start: Unsafe${Mutable}Pointer<Element>?,
+    count: Int
+  ) {
+    _position = start
+    self.count = count
+  }
+
   /// Creates a new buffer pointer over the specified number of contiguous
   /// instances beginning at the given pointer.
   ///
@@ -413,8 +422,7 @@ extension Unsafe${Mutable}BufferPointer {
     _debugPrecondition(
       count == 0 || start != nil,
       "Unsafe${Mutable}BufferPointer has a nil start and nonzero count")
-    _position = start
-    self.count = _assumeNonNegative(count)
+    self.init(_uncheckedStart: start, count: _assumeNonNegative(count))
   }
 
   @inlinable // unsafe-performance

--- a/stdlib/public/core/UnsafeBufferPointer.swift.gyb
+++ b/stdlib/public/core/UnsafeBufferPointer.swift.gyb
@@ -394,6 +394,8 @@ extension Unsafe${Mutable}BufferPointer: ${Mutable}Collection, RandomAccessColle
 }
 
 extension Unsafe${Mutable}BufferPointer {
+  // This works around _debugPrecondition() impacting the performance of
+  // optimized code. (rdar://72246338)
   @_alwaysEmitIntoClient
   internal init(
     @_nonEphemeral _uncheckedStart start: Unsafe${Mutable}Pointer<Element>?,

